### PR TITLE
Feature/resize project manager sidebar

### PR DESCRIPTION
**Pull request by YemSalat**
A feature proposal for a resizable project sidebar.

This is a rather 'quick' implementation, it only allows resizing if `autoHideProject` setting is off. I did run some manual tests, but would be good to have some automatic testing setup.

PS Some more ideas for future: could save the last set sidebar width to keep it between sessions.

### CODE REVIEW

File: js/ui/projectManager.js [hunk 1]
```diff
@@ -129,9 +129,18 @@ define([
     this.expanded = {};
     this.project = null;
     this.projectFile = null;
+
     if (element) {
       this.setElement(element);
+
+      // use host document object to bind resize events
+      if (document && !document.boundSidebarRersizeEvents) {
+        setupSidebarResize(element, document);
```
Comment (YemSalat) [inline js/ui/projectManager.js]: Not sure if best approach here, let me know if needs refactoring

File: js/ui/projectManager.js [hunk 2]
```diff
@@ -129,9 +129,18 @@ define([
     this.expanded = {};
     this.project = null;
     this.projectFile = null;
+
     if (element) {
       this.setElement(element);
+
+      // use host document object to bind resize events
+      if (document && !document.boundSidebarRersizeEvents) {
```
Comment (thomaswilburn) [inline js/ui/projectManager.js]: I'm not wild about keeping the flag on the document (and it would need to be spelled correctly). I'm pretty sure we only call the constructor for the project manager once anyway, so we shouldn't need to guard against multiple bindings.

Comment (YemSalat) [inline js/ui/projectManager.js]: Yep, cool, added it only because was not sure if it does ever get instantiated more then once.
Will remove the flag (sorry about the typo as well).

File: js/ui/projectManager.js [hunk 3]
```diff
@@ -505,12 +515,66 @@ define([
...
+  var MAX_WIDTH = 500;
+
+  function setupSidebarResize(element, document) {
+    // Sidebar resizing
+    var isResizing = false;
+    // make sure width is not set in element `style`
+    element.style.width = null;
+
+    var startSidebarResize = function (evt) {
+      if (evt.target.hasClass('project-resizer')) {
```
Comment (thomaswilburn) [inline js/ui/projectManager.js]: You keep referencing this element, but I'm not aware of having one. Did you forget to commit the index.html changes that added it?

File: css/project.less [hunk 4]
```diff
@@ -21,27 +21,31 @@
     transition: width .3s ease-in-out .3s, border-width .3s ease-in-out .3s, padding .3s linear .3s;
   }
 
+  &._resize {
```
Comment (thomaswilburn) [inline css/project.less]: We don't typically use _ prefixes on classes or variables in Caret (the sandboxing means we can't have real public/private API separation, so I'm pretty cavalier about it). You can just call this `resizing` instead.

File: js/ui/projectManager.js [hunk 5]
```diff
@@ -505,12 +515,66 @@ define([
...
+      //Ace doesn't know about non-window resize events
+      //moving the panel will screw up its dimensions
+      setTimeout(function() {
+        editor.resize();
+      }, 100);
+    };
+
+    document.addEventListener('mousedown', startSidebarResize);
+    document.addEventListener('mousemove', doSidebarResize);
+    document.addEventListener('mouseup', stopSidebarResize);
```
Comment (thomaswilburn) [inline js/ui/projectManager.js]: I think it might be nice to add and remove these listeners on demand, instead of having them available all the time. I know you have the early exit based on `isResizing` in each one, but it's not a bad idea to just cut the function call over head altogether.

File: js/ui/projectManager.js [hunk 6]
```diff
@@ -505,12 +515,66 @@ define([
...
+
+        element.addClass('_resize');
+        isResizing = true;
+      }
+    };
+
+    var doSidebarResize = function (evt) {
+      if (!isResizing) return;
+
+      var mouseX = evt.pageX;
```
Comment (thomaswilburn) [inline js/ui/projectManager.js]: Is pageX reliable for this purpose? I can never remember whether pageX or clientX is more useful for mouse movement.

File: css/project.less [hunk 7]
```diff
@@ -148,6 +152,32 @@
...
\ No newline at end of file
+  position: absolute;
+  height: 100%;
+  right: 0;
+  top: 0;
+
+  width: 16px;
+
+  .project-resizer {
+    &:extend(.project-resizer-wrapper);
```
Comment (thomaswilburn) [inline css/project.less]: I haven't pulled this yet, so forgive me if it's a stupid question, but would it be possible to use a pseudo-element, instead of needing the handle and the wrapper both?

File: css/project.less (hunk 4 above)
Comment (YemSalat) [inline css/project.less]: Yep no worries, will change.
PS here it is not for private/public separation, I usually use `_` prefixes in css for classes that represent states, e.g. `_active`, `_open`, etc. But definitely better to be consistent, so will rename to `resizing`

File: js/ui/projectManager.js (hunk 3 above)
Comment (YemSalat) [inline js/ui/projectManager.js]: Here it is: https://github.com/thomaswilburn/Caret/pull/530/commits/d6d250c899a884b74ec4da4310642a51444dfbd9#diff-4d1f6d0095e22bb74e2b8a36ff770856R30 (sorry it there are a bit too many changes with all the whitespace removed, I put it in a separate commit: https://github.com/thomaswilburn/Caret/pull/530/commits/d6d250c899a884b74ec4da4310642a51444dfbd9)

Comment (thomaswilburn) [inline js/ui/projectManager.js]: That's fine. Before this gets merged, I'll want you to squash the commits into one anyway.

File: js/ui/projectManager.js (hunk 6 above)
Comment (YemSalat) [inline js/ui/projectManager.js]: Yep, `pageX` is relative to the document, while `clientX` is relative to `viewport` (same thing in caret currently, but if it ever starts to matter - `pageX` will be more correct usage)

File: css/project.less (hunk 7 above)
Comment (YemSalat) [inline css/project.less]: I know that it would be better to handle this with pseudo elements from the css point of view, but unfortunately we need to track the click on the bar and pseudo-elements do not allow this.
The outside wrapper here is just to give the sidebar resizer some delay before appearing, so that the cursor does not flicker when the user is moving the mouse between the editor area and the sidebar (cursor changes to `resize-col` when over resizer).

PS Just thought it would also be nice to add `pointer-events: none` there (will update pr)

File: js/ui/projectManager.js (hunk 5 above)
Comment (YemSalat) [inline js/ui/projectManager.js]: Yep, will do

**Comment by YemSalat** [comment]
@thomaswilburn refactored based on feedback. Made it so that the event listeners are removed after the sidebar is resized. Also tweaked a couple other things that you mentioned above.

**Comment by YemSalat** [comment]
Tweaked a bit more to use the convinience `on / off` methods instead of `add/removeEventListener` (had to tweak dom2.js to achieve this)

File: js/util/dom2.js [hunk 8]
```diff
@@ -19,10 +19,11 @@ var el = Element.prototype;
 var doc = Document.prototype;
 var frag = DocumentFragment.prototype;
 var win = Window.prototype;
+var eventTarget = EventTarget.prototype;
```
Comment (YemSalat) [inline js/util/dom2.js]: using `EventTarget` prototype instead of window / element, because they both inherit from EventTarget.

**Comment by YemSalat** [comment]
Squashed the commits, did some more manual testing and it seems good to me.
Please let me know if you want me to tweak anything else.
